File: backend/stage_config.py
"""
Stage Configuration for Backend

Centralized configuration for all deliberation stages.
This ensures consistency between backend event emission and frontend handling.

This config should match frontend/src/stageConfig.js
"""

from typing import Dict, List, Any
from dataclasses import dataclass, field


@dataclass
class SubStageConfig:
    """Configuration for a sub-stage (used in complex stages)"""
    name: str
    label: str
    event_start: str
    event_complete: str


@dataclass
class StageConfig:
    """Configuration for a deliberation stage"""
    name: str
    label: str
    message_field: str  # Which field in assistant message to update
    loading_message: str
    event_start: str
    event_complete: str
    is_complex: bool = False
    sub_stages: List[SubStageConfig] = field(default_factory=list)
    has_metadata: bool = False
    metadata_field: str = "metadata"


# Stage definitions - MUST match frontend/src/stageConfig.js
STAGES = [
    StageConfig(
        name="stage1",
        label="Stage 1: Individual Responses",
        message_field="stage1",
        loading_message="Collecting individual responses...",
        event_start="stage1_start",
        event_complete="stage1_complete"
    ),
    StageConfig(
        name="stage1_5",
        label="Stage 1.5: Cross-Interrogation",
        message_field="stage1_5",
        loading_message="Running cross-interrogation...",
        event_start="stage1_5_questions_start",  # First sub-stage start
        event_complete="stage1_5_answers_complete",  # Last sub-stage complete
        is_complex=True,
        sub_stages=[
            SubStageConfig(
                name="questions",
                label="Generating Questions",
                event_start="stage1_5_questions_start",
                event_complete="stage1_5_questions_complete"
            ),
            SubStageConfig(
                name="answers",
                label="Collecting Answers",
                event_start="stage1_5_answers_start",
                event_complete="stage1_5_answers_complete"
            )
        ]
    ),
    StageConfig(
        name="stage2",
        label="Stage 2: Peer Rankings",
        message_field="stage2",
        loading_message="Collecting peer rankings...",
        event_start="stage2_start",
        event_complete="stage2_complete",
        has_metadata=True,
        metadata_field="metadata"
    ),
    StageConfig(
        name="stage3",
        label="Stage 3: Final Synthesis",
        message_field="stage3",
        loading_message="Synthesizing final answer...",
        event_start="stage3_start",
        event_complete="stage3_complete"
    )
]

# Special event names (non-stage events)
SPECIAL_EVENTS = {
    "STREAM_INIT": "stream_init",
    "RESUME_INIT": "resume_init",
    "COMPLETE": "complete",
    "ERROR": "error",
    "HEARTBEAT": "heartbeat",
    "RECONNECTED": "reconnected",
    "AUTH_EXPIRED": "auth_expired",
    "TITLE_COMPLETE": "title_complete"
}


def get_stage_config(stage_name: str) -> StageConfig:
    """Get stage configuration by name"""
    for stage in STAGES:
        if stage.name == stage_name:
            return stage
    return None


def get_all_event_types() -> List[str]:
    """Get all event types for validation"""
    stage_events = []
    for stage in STAGES:
        if stage.is_complex:
            for sub_stage in stage.sub_stages:
                stage_events.append(sub_stage.event_start)
                stage_events.append(sub_stage.event_complete)
        else:
            stage_events.append(stage.event_start)
            stage_events.append(stage.event_complete)

    special_events = list(SPECIAL_EVENTS.values())
    return stage_events + special_events


def get_stage_names() -> List[str]:
    """Get ordered list of stage names"""
    return [stage.name for stage in STAGES]


def validate_stage_sequence(last_stage: str) -> List[str]:
    """
    Get remaining stages after a given stage.
    Used for stream resumption.

    Args:
        last_stage: Name of last completed stage

    Returns:
        List of remaining stage names
    """
    stage_names = get_stage_names()
    try:
        last_index = stage_names.index(last_stage)
        return stage_names[last_index + 1:]
    except ValueError:
        raise ValueError(f"Invalid stage name: {last_stage}")


# Example usage for generating events in conversations.py:
#
# from .stage_config import get_stage_config, SPECIAL_EVENTS
#
# stage = get_stage_config("stage1")
# yield send_event(stage.event_start, stage='stage1')
# # ... do work ...
# yield send_event(stage.event_complete, data, 'stage1')

File: frontend/src/stageConfig.js
/**
 * Stage Configuration
 *
 * Centralized configuration for all deliberation stages.
 * This defines the contract between backend events and frontend rendering.
 *
 * Adding a new stage:
 * 1. Add entry to STAGES array with configuration
 * 2. Backend should emit events matching the pattern defined here
 * 3. Frontend will automatically handle the events and update UI
 */

export const STAGES = [
  {
    name: 'stage1',
    label: 'Stage 1: Individual Responses',
    messageField: 'stage1', // Which field in assistant message to update
    loadingMessage: 'Collecting individual responses...',
    events: {
      start: 'stage1_start',
      complete: 'stage1_complete'
    }
  },
  {
    name: 'stage1_5',
    label: 'Stage 1.5: Cross-Interrogation',
    messageField: 'stage1_5',
    loadingMessage: 'Running cross-interrogation...',
    // Complex stage with sub-events
    isComplex: true,
    subStages: [
      {
        name: 'questions',
        label: 'Generating Questions',
        events: {
          start: 'stage1_5_questions_start',
          complete: 'stage1_5_questions_complete'
        },
        // Custom data handler for sub-stage
        dataHandler: (event, currentData) => ({
          ...(currentData || {}),
          questions: event.data
        })
      },
      {
        name: 'answers',
        label: 'Collecting Answers',
        events: {
          start: 'stage1_5_answers_start',
          complete: 'stage1_5_answers_complete'
        },
        // Custom data handler that also captures label_to_model
        dataHandler: (event, currentData) => ({
          ...(currentData || {}),
          answers: event.data,
          label_to_model: event.label_to_model
        })
      }
    ]
  },
  {
    name: 'stage2',
    label: 'Stage 2: Peer Rankings',
    messageField: 'stage2',
    loadingMessage: 'Collecting peer rankings...',
    events: {
      start: 'stage2_start',
      complete: 'stage2_complete'
    },
    // Stage 2 includes metadata
    hasMetadata: true,
    metadataField: 'metadata'
  },
  {
    name: 'stage3',
    label: 'Stage 3: Final Synthesis',
    messageField: 'stage3',
    loadingMessage: 'Synthesizing final answer...',
    events: {
      start: 'stage3_start',
      complete: 'stage3_complete'
    }
  }
];

// Special non-stage events
export const SPECIAL_EVENTS = {
  STREAM_INIT: 'stream_init',
  RESUME_INIT: 'resume_init',
  COMPLETE: 'complete',
  ERROR: 'error',
  HEARTBEAT: 'heartbeat',
  RECONNECTED: 'reconnected',
  AUTH_EXPIRED: 'auth_expired',
  TITLE_COMPLETE: 'title_complete'
};

// Workflow event patterns and configuration
export const WORKFLOW_EVENTS = {
  STREAM_INIT: 'stream_init',
  SUPERSTEP_MAP_START: /^superstep_(\w+)_map_start$/,
  SUPERSTEP_MAP_COMPLETE: /^superstep_(\w+)_map_complete$/,
  SUPERSTEP_MIDDLEWARE_COMPLETE: /^superstep_(\w+)_middleware_complete$/,
  SUPERSTEP_REDUCE_START: /^superstep_(\w+)_reduce_start$/,
  SUPERSTEP_REDUCE_COMPLETE: /^superstep_(\w+)_reduce_complete$/,
  COMPLETE: 'complete'
};

/**
 * Check if event type is a workflow event
 */
export function isWorkflowEvent(eventType) {
  // Check exact matches
  if (eventType === WORKFLOW_EVENTS.STREAM_INIT || eventType === WORKFLOW_EVENTS.COMPLETE) {
    return true;
  }

  // Check pattern matches
  return Object.values(WORKFLOW_EVENTS).some(pattern => {
    if (pattern instanceof RegExp) {
      return pattern.test(eventType);
    }
    return false;
  });
}

/**
 * Parse workflow event to extract step_id and phase
 */
export function parseWorkflowEvent(eventType) {
  for (const [key, pattern] of Object.entries(WORKFLOW_EVENTS)) {
    if (pattern instanceof RegExp) {
      const match = eventType.match(pattern);
      if (match) {
        return {
          type: key,
          stepId: match[1],
          eventType
        };
      }
    }
  }

  return null;
}

/**
 * Get stage configuration by name
 */
export function getStageConfig(stageName) {
  return STAGES.find(stage => stage.name === stageName);
}

/**
 * Get all event types (for validation/debugging)
 */
export function getAllEventTypes() {
  const stageEvents = STAGES.flatMap(stage => {
    if (stage.isComplex) {
      return stage.subStages.flatMap(sub =>
        Object.values(sub.events)
      );
    }
    return Object.values(stage.events);
  });

  const specialEvents = Object.values(SPECIAL_EVENTS);

  return [...stageEvents, ...specialEvents];
}

/**
 * Find which stage handles a given event type
 */
export function getStageForEvent(eventType) {
  for (const stage of STAGES) {
    if (stage.isComplex) {
      // Check sub-stages
      for (const subStage of stage.subStages) {
        if (Object.values(subStage.events).includes(eventType)) {
          return { stage, subStage };
        }
      }
    } else {
      // Check main stage events
      if (Object.values(stage.events).includes(eventType)) {
        return { stage };
      }
    }
  }
  return null;
}

/**
 * Check if event type is a special (non-stage) event
 */
export function isSpecialEvent(eventType) {
  return Object.values(SPECIAL_EVENTS).includes(eventType);
}

/**
 * Get ordered list of stage names (for stage sequence)
 */
export function getStageNames() {
  return STAGES.map(s => s.name);
}
